substrate/client/network/sync/src/strategy.rs

```diff
@@ -0,0 +1,463 @@
+	SendWarpProofRequest { peer_id: PeerId, request: WarpProofRequest<B> },
+	/// Peer misbehaved. Disconnect, report it and cancel any requests to it.
+	DropPeer(BadPeer),
+	/// Import blocks.
+	ImportBlocks { origin: BlockOrigin, blocks: Vec<IncomingBlock<B>> },
+	/// Import justifications.
+	ImportJustifications {
+		peer_id: PeerId,
+		hash: B::Hash,
+		number: NumberFor<B>,
+		justifications: Justifications,
+	},
+	/// Syncing strategy has finished.
+	Finished,
+}
+
+/// Proxy to specific syncing strategies.
+pub enum SyncingStrategy<B: BlockT, Client> {
+	WarpSyncStrategy(WarpSync<B, Client>),
+	StateSyncStrategy(StateStrategy<B, Client>),
+	ChainSyncStrategy(ChainSync<B, Client>),
+}
+
+impl<B: BlockT, Client> SyncingStrategy<B, Client>
+where
+	B: BlockT,
+	Client: HeaderBackend<B>
+		+ BlockBackend<B>
+		+ HeaderMetadata<B, Error = sp_blockchain::Error>
+		+ ProofProvider<B>
+		+ Send
+		+ Sync
+		+ 'static,
+{
+	/// Initialize a new syncing startegy.
+	pub fn new(
+		config: &SyncingConfig,
+		client: Arc<Client>,
+		warp_sync_config: Option<WarpSyncConfig<B>>,
+	) -> Result<Self, ClientError> {
+		if let SyncMode::Warp = config.mode {
+			let warp_sync_config = warp_sync_config
+				.expect("Warp sync configuration must be supplied in warp sync mode.");
+			Ok(Self::WarpSyncStrategy(WarpSync::new(client.clone(), warp_sync_config)))
+		} else {
+			Ok(Self::ChainSyncStrategy(ChainSync::new(
+				chain_sync_mode(config.mode),
+				client.clone(),
+				config.max_parallel_downloads,
+				config.max_blocks_per_request,
+			)?))
+		}
+	}
+
+	/// Notify that a new peer has connected.
+	pub fn new_peer(&mut self, peer_id: PeerId, best_hash: B::Hash, best_number: NumberFor<B>) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(strategy) =>
+				strategy.new_peer(peer_id, best_hash, best_number),
+			SyncingStrategy::StateSyncStrategy(strategy) =>
+				strategy.new_peer(peer_id, best_hash, best_number),
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.new_peer(peer_id, best_hash, best_number),
+		}
+	}
+
+	/// Notify that a peer has disconnected.
+	pub fn peer_disconnected(&mut self, peer_id: &PeerId) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(strategy) => strategy.peer_disconnected(peer_id),
+			SyncingStrategy::StateSyncStrategy(strategy) => strategy.peer_disconnected(peer_id),
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.peer_disconnected(peer_id),
+		}
+	}
+
+	/// Submit a validated block announcement.
+	///
+	/// Returns new best hash & best number of the peer if they are updated.
+	pub fn on_validated_block_announce(
+		&mut self,
+		is_best: bool,
+		peer_id: PeerId,
+		announce: &BlockAnnounce<B::Header>,
+	) -> Option<(B::Hash, NumberFor<B>)> {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) =>
+				Some((announce.header.hash(), *announce.header.number())),
+			SyncingStrategy::StateSyncStrategy(_) =>
+				Some((announce.header.hash(), *announce.header.number())),
```

Comment:
Block announce validation happens in `SyncingEngine`, so it's already validated. The only exception is `ChainSync` where we do not update best hash & number if we know that the peer is in the ancestry search state.

Otherwise, updating best hash & number upon processing a validated block announcement seems not worse than updating it upon receiving a handshake, which also says nothing about what fork the peer is at.

Reply:
Fair enough